### DISCUSSION TITLE
Glyph Alternates in OpenType (-ish?) fonts

### DISCUSSION THREAD

**Discussion by kellyo-cricut** [harfbuzz/harfbuzz#4305] [Q&A] [answered]
I’m trying to query for alternates in Zapfino, a macOS system font, using harfbuzz 7.2.0. Looking at the font data (using [fontdrop.io](http://fontdrop.io/)) doesn’t detect any OpenType features, but it does show a GSUB table that has a few OT-looking features for glyph alternates like aalt, calt, etc. When I query for font features using `hb_ot_layout_table_get_feature_tags` does not find any features. Calling `hb_ot_layout_collect_lookups` not specifying any scripts, languages, or features does not return any indices. Calling `hb_ot_layout_lookup_get_glyph_alternates` with indices found in the GSUB table does not return any alternates for a glyph id that has mappings inside the GSUB table.
 
Most of the text/font world is quite new to me, so apologies if I’m missing something obvious here. Is this a case of some oddly formed font data? I’m able to shape using alternate glyphs by id using the solution [from this discussion](https://github.com/harfbuzz/harfbuzz/discussions/4170), but I’m at a loss as to how to retrieve the IDs of glyphs that are mapped in one of the GSUB features. TextEdit offers alternates for those glyphs, so I’m hoping there is a way to query for those with harfbuzz. Thanks!

**Comment by behdad** [comment]
The version of Zapfino I have access to doesn't have GSUB at all. So I can't help. Maybe show us a ttx dump of the GSUB table?

**Reply by kellyo-cricut** [reply]
Ah- I'm not seeing a GSUB table using that tool:
<img width="655" alt="image" src="https://github.com/harfbuzz/harfbuzz/assets/57497291/2815d30a-89a8-468c-b032-5b12d79c9c3e">
I'm unsure of what fontdrop.io is doing, but the same file is suggesting that there is a GSUB table with mappings:
<img width="747" alt="image" src="https://github.com/harfbuzz/harfbuzz/assets/57497291/dbc2b1e6-1f65-4e4b-be57-104c644d2483">
Im assuming ttx is correct here. Knowing that, is there another way to query for alternate glyphs? I'm trying to get the same alternates that TextEdit does here:
<img width="589" alt="image" src="https://github.com/harfbuzz/harfbuzz/assets/57497291/dabefaf2-359d-41e7-8e72-761855c63c64">

**Comment by behdad** [comment, marked as answer]
> I'm unsure of what fontdrop.io is doing, but the same file is suggesting that there is a GSUB table with mappings:

I don't get any OT features in fontdrop for the font I have.

> Im assuming ttx is correct here. Knowing that, is there another way to query for alternate glyphs? I'm trying to get the same alternates that TextEdit does here:

I'm afraid we don't implement that API for AAT fonts. I'm not even sure which table stores those. Probably the `Zapf` table? @ebraminio

**Reply by kellyo-cricut** [reply]
> I don't get any OT features in fontdrop for the font I have.

I don't see any OT features reported there either. After looking at the ttx dump I'm very confused how and why GSUB is shown there 🤷

**Reply by behdad** [reply]
Where does it show GSUB?

**Reply by behdad** [reply]
Ah, I see. In Data.

**Reply by kellyo-cricut** [reply]
Looking at some of the dumps from ttx, it looks like that mapping data is in the `morx` table. Is  that still part of AAT that harfbuzz doesn't search?

**Reply by behdad** [reply]
HarfBuzz applies morx but doesn't query it no. In general that's a very tricky problem since morx contains state-machines, not regular tables.

**Reply by khaledhosny** [reply]
Looks like Zapf table would have such data indeed, see https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6Zapf.html (Structures For Groups of Glyphs section)
